cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -313,54 +257,97 @@ struct rle_stream {
     int const warp_lane      = t % cudf::detail::warp_size;
 
-    __shared__ int run_start;
-    __shared__ int num_runs;
-    __shared__ int values_processed;
-    if (!t) {
-      // carryover from the last call.
-      thrust::tie(run_start, num_runs) = get_run_batch();
-      values_processed                 = 0;
+    __shared__ int values_processed_shared;
+    __shared__ int decode_index_shared;
+    __shared__ int fill_index_shared;
+    if (t == 0) {
+      values_processed_shared = 0;
+      decode_index_shared     = decode_index;
+      fill_index_shared       = fill_index;
     }
+
     __syncthreads();
 
+    fill_index = fill_index_shared;
+
     do {
-      // warp 0 reads ahead and generates batches of runs to be decoded by remaining warps.
-      if (!warp_id) {
+      // protect against threads advancing past the end of this loop
+      // and updating shared variables.
+      __syncthreads();
+
+      // warp 0 reads ahead and fills `runs` array to be decoded by remaining warps.
+      if (warp_id == 0) {
         // fill the next set of runs. fill_runs will generally be the bottleneck for any
         // kernel that uses an rle_stream.
-        if (warp_lane == 0) { fill_run_batch(output_count); }
-      }
-      // remaining warps decode the runs
-      else if (warp_decode_id < num_runs) {
-        // each warp handles 1 run, regardless of size.
-        // TODO: having each warp handle exactly 32 values would be ideal. as an example, the
-        // repetition levels for one of the list benchmarks decodes in ~3ms total, while the
-        // definition levels take ~11ms - the difference is entirely due to long runs in the
-        // definition levels.
-        auto& run  = runs[rolling_index<run_buffer_size>(run_start + warp_decode_id)];
-        auto batch = run.next_batch(output + run.output_pos,
-                                    min(run.remaining, (output_count - run.output_pos)));
-        batch.decode(end, level_bits, warp_lane, warp_decode_id);
-        // last warp updates total values processed
-        if (warp_lane == 0 && warp_decode_id == num_runs - 1) {
-          values_processed = run.output_pos + batch.size;
+        if (warp_lane == 0) {
+          fill_run_batch();
+          if (decode_index == -1) {
+            // first time, set it to the beginning of the buffer (rolled)
+            decode_index        = 0;
+            decode_index_shared = decode_index;
+          }
+          fill_index_shared = fill_index;
         }
       }
-      __syncthreads();
-
-      // if we haven't run out of space, retrieve the next batch. otherwise leave it for the next
-      // call.
-      if (!t && values_processed < output_count) {
-        thrust::tie(run_start, num_runs) = get_run_batch();
+      // remaining warps decode the runs, starting on the second iteration of this
+      // do/while loop (decode_index == -1 means "first iteration", so we should skip decoding)
```

Comment:
Maybe worth mentioning that the first-iteration is literally the first iteration across all calls to the function.  To highlight the "keep the pipeline" full angle.